[Cocoa] [Downlevels] Out of bounds optical sizing is rendered as the default value instead of clamping

litherum opened the pull request:
#### e9e9b4699c26d5e69ffeafd259edcc7e203479fb
<pre>
[Cocoa] [Downlevels] Out of bounds optical sizing is rendered as the default value instead of clamping
<a href="https://bugs.webkit.org/show_bug.cgi?id=258764">https://bugs.webkit.org/show_bug.cgi?id=258764</a>
rdar://111535841

Reviewed by Cameron McCormack.

On pre-2023-aligned Apple OSes, Core Text&apos;s behavior for optical sizing is that
out of bounds values are ignored, and default optical sizing is used instead of
clamping the specified value. We can work around this by clamping the value
ourself in WebKit, but this is somewhat unfortunate because we only know the
acceptable range after creating a CTFont, but we&apos;d like to specify the values
during CTFont creation time.

This patch solves the problem by doing a 2-phase creation: Create a CTFont,
inspect it, determine if it needs a second phase, and if it does, modify the
CTFont to create a second, derived, CTFont. We already have some support for
this due to the TrueType GX weight/width/slope scales being different than
the CSS scales - but we don&apos;t know if a font is a TrueType GX font until we&apos;ve
created the CTFont object. This patch extends that existing 2-phase creation to
also recreate the font if the optical sizing ended up wrong.

* LayoutTests/fast/text/variations/optical-sizing-out-of-bounds-expected.html: Added.
* LayoutTests/fast/text/variations/optical-sizing-out-of-bounds.html: Added.
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
(WebCore::OpticalSizeRange::isInRange const):
(WebCore::opticalSizeRange):
(WebCore::UnrealizedCoreTextFont::rebuildReason const):
(WebCore::UnrealizedCoreTextFont::realize const):
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:
(WebCore::UnrealizedCoreTextFont::RebuildReason::hasEffect const):

Canonical link: <a href="https://commits.webkit.org/266000@main">https://commits.webkit.org/266000@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f72cb6a493240c4914fa60e3d53698ac9e247abf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12352 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12998 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14095 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11881 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12408 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15113 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12705 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14564 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12516 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13282 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10449 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14517 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10565 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11194 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18299 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10506 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11359 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14541 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/11690 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11848 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9780 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12414 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11074 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11065 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3259 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3077 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15404 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/12767 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11701 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3040 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:
```diff
@@ -96,6 +107,7 @@ class UnrealizedCoreTextFont {
     float m_width { 0 };
     float m_slope { 0 };
     CGFloat m_size { 0 };
+    std::optional<CGFloat> m_sizeForOpticalSizing;
```
This should be deleted

vitorroriz commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -397,3 +397,10 @@
+    || (PLATFORM(APPLETV) && __TV_OS_VERSION_MAX_ALLOWED < 170000) \
+    || (PLATFORM(WATCHOS) && __WATCH_OS_VERSION_MAX_ALLOWED < 100000)
+#define USE_CORE_TEXT_OPTICAL_SIZING_RANGE_CLAMPING_WORKAROUND 1
+#endif
```
I'm not sure if we prefer having one guard for each flag, but this uses the same guard as `USE_VISION_CPU_ONLY_PROPERTY 1` and `USE_METAL_ARGUMENT_ACCESS_ENUMS`.

litherum replied:
Historically, we've used one guard for each flag.

vitorroriz commented on Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
```diff
@@ -334,6 +335,114 @@ void UnrealizedCoreTextFont::modifyFromContext(const FontDescription& fontDescri
+        if (iterator != m_variationSettings.end())
+            rebuildReason.outOfRangeVariations = std::clamp(iterator->value(), static_cast<float>(range->minimum), static_cast<float>(range->maximum));
+    }
+#endif
```
can't we guard this fragment against `FontDescription::opticalSizing()` such that we avoid querying CT if optical sizing is ::Disabled?

litherum replied:
That's what `OpticalSizingTypes::None` does. If optical sizing is disabled, then `m_opticalSizingType` is set to `OpticalSizingTypes::None`.

heycam commented on Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
```diff
@@ -356,48 +430,89 @@ RetainPtr<CTFontRef> UnrealizedCoreTextFont::realize() const
+            slope = denormalizeSlope(m_slope);
+        }
+
+        {
```
Nit: this block doesn't seem to be needed.

litherum replied:
I think I'd rather keep it because it gives the reader a visual grouping of the various pieces of this function.

litherum committed:
[Cocoa] [Downlevels] Out of bounds optical sizing is rendered as the default value instead of clamping
https://bugs.webkit.org/show_bug.cgi?id=258764
rdar://111535841

Reviewed by Cameron McCormack.

On pre-2023-aligned Apple OSes, Core Text's behavior for optical sizing is that
out of bounds values are ignored, and default optical sizing is used instead of
clamping the specified value. We can work around this by clamping the value
ourself in WebKit, but this is somewhat unfortunate because we only know the
acceptable range after creating a CTFont, but we'd like to specify the values
during CTFont creation time.

This patch solves the problem by doing a 2-phase creation: Create a CTFont,
inspect it, determine if it needs a second phase, and if it does, modify the
CTFont to create a second, derived, CTFont. We already have some support for
this due to the TrueType GX weight/width/slope scales being different than
the CSS scales - but we don't know if a font is a TrueType GX font until we've
created the CTFont object. This patch extends that existing 2-phase creation to
also recreate the font if the optical sizing ended up wrong.

* LayoutTests/fast/text/variations/optical-sizing-out-of-bounds-expected.html: Added.
* LayoutTests/fast/text/variations/optical-sizing-out-of-bounds.html: Added.
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
(WebCore::OpticalSizeRange::isInRange const):
(WebCore::opticalSizeRange):
(WebCore::UnrealizedCoreTextFont::rebuildReason const):
(WebCore::UnrealizedCoreTextFont::realize const):
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:
(WebCore::UnrealizedCoreTextFont::RebuildReason::hasEffect const):

Canonical link: https://commits.webkit.org/266000@main
